File: DSA/50_Days_50_Questions/713-subarray-product-less-than-k/subarray-product-less-than-k.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int n = nums.size();

        if(k<=1){
            return 0;
        }
        int l =0, ans=0, prd=1;

        for(int r=0; r<n; r++){
            prd*=nums[r];

            while(prd>=k){
                prd/=nums[l];
                l++;
            }

            ans += (r-l)+1;
        }
        return ans;
    }
};

File: DSA/50_Days_50_Questions/713-subarray-product-less-than-k/subarray-product-less-than-k.py
class Solution:
    def numSubarrayProductLessThanK(self, nums, k):
        if k <= 1:
            return 0

        left = 0
        product = 1
        answer = 0

        for right in range(len(nums)):
            product *= nums[right]

            while product >= k:
                product //= nums[left]
                left += 1

            answer += right - left + 1

        return answer